server/src/main/java/org/opensearch/bootstrap/Bootstrap.java

```diff
@@ -336,8 +347,10 @@ private static Environment createEnvironment(
     }
 
     private void start() throws NodeValidationException {
-        node.start();
+        // keepAliveThread should start first than node to ensure the cluster can spin up successfully in edge cases:
+        // https://github.com/opensearch-project/OpenSearch/issues/14791
         keepAliveThread.start();
+        node.start();
```

Comment:
Naively speaking, if `node.start()` doesn't succeeded, then there is nothing to keep alive and the JVM shutting down seems like the right thing to do. It seems like this change could lead to a partially-started or not-at-all-started node running indefinitely because the keepAliveThread will just keep it alive in some sort of zombie state after `node.start()` failed. What am I missing?

Reply:
Naively responding, I've no objection to the JVM shutting down, but having spent way too many hours trying to figure out *why* the JVM shut down in various situations, I'd be really happy to have at least one thread faithfully logging whatever issues caused it to shut down, before shutting itself down.

Reply:
Not all `node.start()` failing means the JVM should quit, like the case in the corresponding issue, if we're able to keep JVM running, user can fix this issue simply by changing the cluster settings. So the fix is to provide user the capability to interact with the running cluster(even partially-started).
And for those cases `node.start()` failing and the cluster is not available at all case, this fix doesn't have real impact because in the end user needs to check error/fatal logs to fix the root cause. This change doesn't add any blocking to user but w/o this change, the first case users are blocked. 
So in all I think this is a positive enhancement to user.

Reply:
> like the case in the corresponding issue, if we're able to keep JVM running, user can fix this issue simply by changing the cluster settings.

This seems like a really blunt change for an extremely specific case. In general, quitting the JVM and letting it restart is preferable to continuing to run in a partially-started state. Prior to this change, a transient failure in `node.start()` might automatically recovery because the JVM would quit and whatever is monitoring the process would restart it. Now there is a risk (in my opinion) that the process will continue running in a non-functional partially-started state that will require human intervention to resolve. How do we know that is not a risk here?

Reply:
> but having spent way too many hours trying to figure out why the JVM shut down in various situations, I'd be really happy to have at least one thread faithfully logging whatever issues caused it to shut down, before shutting itself down.

@dbwiddis Not logging why the JVM shutdown is clearly a problem. However, I'd argue this change might make things considerably worse. Instead of knowing there's a problem (because the process restarted) we'll instead have a partially-started node running in a crippled state with no indication that something went wrong during startup.

Reply:
I'm on the fence here. I see benefits to both sides, and I'm not really an expert in either, so I'll defer to @andrross and @zane-neo here.

Reply:
> might automatically recover because the JVM would quit and whatever is monitoring the process would restart it
My thought on this are:
1. A partially-started cluster may not block user's operation based on the specific usage of the user performs.
2. If a partially-started cluster have issue during usage, user can find issues through their monitor, and in the end user need manual intervention which is same with user found JVM quitting. If your worries are about the data loss during the time range, I'd say they both causing data loss.
3. Currently I only see the observability plugin throwing exception during the startup and I'm not sure if this a good practice. Do you think we can involve people from observability to discuss this?

Also, I'm open to add more changes to make this to have even better user experience if you have proposals, thanks.

Reply:
> A partially-started cluster may not block user's operation based on the specific usage of the user performs.

The reality is that (1) the node doesn't start up due to condition X. (2) ideally it is possible to fix condition X so that the node can start up. (3) This requires diagnosing condition X.  

When interacting with production clusters, I almost always assume that nodes which don't start up have a disk full issue, because that's almost always the case.  But other than just seeing it over and over, that's not very helpful to fixing the problem.

> If a partially-started cluster have issue during usage, user can find issues through their monitor, and in the end user need manual intervention which is same with user found JVM quitting. 

This is the crux of the discussion here.  The simple fix in this PR doesn't actually do anything to inform the user of the problem, or make any sort of decision on whether a half-started node can answer a REST request saying (Hey! I'm out of disk space!) and providing a way to recover.

As it is, the current status quo is the cluster struggles along without one node, which won't start, but also won't terminate and get replaced because data loss, and always requires manual intervention.  

What can we do for the disk usage situation to let users solve this themselves?

Reply:
> This is the crux of the discussion here. The simple fix in this PR doesn't actually do anything to inform the user of the problem, or make any sort of decision on whether a half-started node can answer a REST request saying (Hey! I'm out of disk space!) and providing a way to recover.

In fact the half-started cluster do provide a way to recover this case, e.g. user can change the cluster settings to update the `cluster.routing.allocation.disk.watermark.flood_stage` value to a higher one for a temporarily fix and then increase the disk space to ensure the cluster working well in next days. But without the fix, the cluster is not able to startup and user need much more manual effort to fix this(delete the observability plugin and start the cluster and then reinstall the plugin)...

Reply:
> In fact the half-started cluster do provide a way to recover this case, e.g. user can change the cluster settings

Right, but this doesn't help if the user doesn't know there's a problem.   A failed node (that still has its data intact) is easy to recognize is a problem.  A half-started node might look healthy-ish.  

And that's also only one possible use case here.

Is there a way to put a timeout on the keepalive thread?   Like, give the user some reasonable time window after startup (5 minutes?) to investigate and fix the problem?

Reply:
> If a partially-started cluster have issue during usage, user can find issues through their monitor
>> ...A half-started node might look healthy-ish.

@dbwiddis @zane-neo This is the crux of the issue. We have _no idea_ how a half-started node will behave. We also have no idea what monitors a user may have configured and whether this half-started node will be seen as healthy or unhealthy. To put it concretely, imagine a problem causes `node.start()` to throw an exception. Previous to this change a deployment to a cluster would get stuck when bouncing the first node because it definitely fail to start up and just get stuck in a restart cycle. After this change, we don't know if that deployment would get stuck and the worst case is that the deployment doesn't have comprehensive enough monitors to detect the degraded nodes and it would cycle through all nodes, potentially resulting in a severe outage.

I'm still in favor of rolling back this change.

Reply:
> I'm still in favor of rolling back this change.

I'm in agreement at this point.

Reply:
I agree in some real edge case `a half-started node are healthy from monitor` might be true, but it's more convincible if you can provide examples on this. But let's discuss on other possible solutions first:  
1. Confirm with observability if it's possible to not throw exception to block the cluster startup, I have created an issue: https://github.com/opensearch-project/observability/issues/1872
2. Change the node start code to add disk usage check and update the state when cluster manager is elected, this needs much more efforts since the state transition during startup is complicated and the disk usage check code might need refactor. This also adds a lot of code review effort, I hope you can help on this if this works out.

I believe the second one would be the long term solution, and the first one could be short term. If either one works out, I don't have objection to rollback this change. WDYT?

Reply:
Solution (1) definitely needs to be fixed, particularly in light of https://github.com/opensearch-project/OpenSearch/issues/16340, and not just in the short term.  (And I just realized now that bug is more serious than I thought it was because of this.)

I agree (2) is the right direction to go generally, but don't see much detail there.

Reply:
>  If either one works out, I don't have objection to rollback this change.

Can we roll it back sooner rather than later?  The unmerged backport/lack of changelog will make every backport PR creation fail until the main and 2.x changelogs sync up.

Reply:
I can create the revert PR now, and I'll look into the solution 2 for a long term solution.

Reply:
Thanks @zane-neo .   Please also consider solution 1 and/or the bug I reported.  One of the two of those issues should be fixed ASAP.

Reply:
@andrross @dblock @dbwiddis Do you think we can conclude that all plugins shouldn't block the OpenSearch from starting up even plugin itself encountering issue? I believe if this is true, we can simple do a try catch check for the [onNodeStarted](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/ClusterPlugin.java#L102) method to avoid this. The pros of this fix is it's simple and effective, and it can prevent this won't happen in other plugins.

Reply:
> Do you think we can conclude that all plugins shouldn't block the OpenSearch from starting up

@zane-neo I don't think this is true generally. For example, if the security plugin can't properly start up it would be safer to fail versus coming up with security disabled.

Reply:
@andrross thanks, I created this PR: https://github.com/opensearch-project/observability/pull/1873 for short term fix, and I will continue the long term fix exploration.